cmd/ingester/app/consumer/rate_limiter_test.go

```diff
@@ -0,0 +1,41 @@
+// Copyright (c) 2018 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package consumer
+
+import (
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestRateLimiter(t *testing.T) {
+	const (
+		creditsPerSecond = 100
+		iterations       = 20
+	)
+	rateLimiter := newRateLimiter(creditsPerSecond)
+	defer rateLimiter.stop()
+
+	var m sync.Mutex
+	counter := 0
+	for i := 0; i < iterations; i++ {
+		rateLimiter.await()
+		m.Lock()
+		counter++
+		m.Unlock()
+	}
+	require.Equal(t, counter, iterations)
```

Comment:
The issue I actually saw was that the rate limiting **did not work** on each iteration, but only as a totality. For example, I had tested that each tick was at least 10 ms apart, and had outright assertion failures because the actual tick came at ~9.64 ms, etc. My guess is the clock is not the same, or the ticker guarantees are much more lax than I had thought.